packages/engine/paima-funnel/src/funnels/midnight/funnel.ts

```diff
@@ -0,0 +1,451 @@
+            this.chainInfo.config.confirmationDepth
+      ) {
+        this.cache.wsBlocks.push(handleGqlWsError(await this.cache.iter.next()).blocks);
+        // ^ Hovers around 5.4ish to 6.6ish second wait per block. Nominal block time is 6s.
+      }
+
+      const nextBlock = this.cache.wsBlocks[0];
+
+      if (height < nextBlock.height) {
+        const oldBlock = await this.fetchBlock(height);
+        if (!oldBlock) {
+          throw new Error(`Saw block ${nextBlock.height}, but old block ${height} doesn't exist!`);
+        }
+        return oldBlock;
+      } else if (height == nextBlock.height) {
+        this.cache.wsBlocks.splice(0, 1);
+        return nextBlock;
+      } else {
+        this.cache.wsBlocks.splice(0, 1);
+      }
+    }
+  }
+
+  private extensionsFromBlock(
+    baseBlockNumber: number,
+    block: CachedBlock
+  ): ChainDataExtensionDatum[] {
+    const result: CdeMidnightContractStateDatum[] = [];
+
+    const contractAddressToExtension = new Map(
+      this.sharedData.extensions.flatMap(ext =>
+        ext.network === this.chainInfo.name &&
+        ext.cdeType === ChainDataExtensionType.MidnightContractState
+          ? [[ext.contractAddress, ext]]
+          : []
+      )
+    );
+
+    for (let tx of block.transactions) {
+      for (let contractCall of tx.contractCalls) {
+        const extension = contractAddressToExtension.get(contractCall.address);
+        if (extension) {
+          // Only deserialize if we actually care.
+          setNetworkId(this.chainInfo.config.networkId);
+          const state = ContractState.deserialize(hexStringToUint8Array(contractCall.state));
+          // We could downcast contractCall to see if operations() contains something useful.
+          // For now let's report on the ledger variable contents.
+          const jsState = state.data.encode();
+
+          result.push({
+            blockNumber: baseBlockNumber,
+            transactionHash: tx.hash,
+
+            cdeDatumType: ChainDataExtensionDatumType.MidnightContractState,
+            cdeName: extension.cdeName,
+            contractAddress: contractCall.address,
+            caip2: this.caip2,
+
+            scheduledPrefix: extension.scheduledPrefix,
+            payload: JSON.stringify(jsState),
+          });
+        }
+      }
+    }
+
+    return result;
+  }
+
+  public override async readPresyncData(
+    args: ReadPresyncDataFrom
+  ): Promise<{ [caip2: string]: PresyncChainData[] | typeof FUNNEL_PRESYNC_FINISHED }> {
+    const baseDataPromise = this.baseFunnel.readPresyncData(args);
+
+    const startingBlock = await this.sharedData.mainNetworkApi.getStartingBlock();
+    if (!startingBlock) {
+      throw new Error("Couldn't get main's network staring block timestamp");
+    }
+
+    const start = this.cache.nextBlockHeight;
+    let block = await this.fetchBlock(this.cache.nextBlockHeight);
+    if (!(block && midnightTimestampToSeconds(block.timestamp) < startingBlock.timestamp)) {
```

Comment:
just personal preference, but isn't this more clear as:

```ts
if (!block || midnightTimestampToSeconds(block.timestamp) >= startingBlock.timestamp) {
```

Although I'm not sure I understand why the block being null implies that it's finished.

Reply:
`block` being null means we hit the head of the Midnight chain. Perhaps the presync should stall for new blocks just like regular sync in that case?